lib/ibandit/german_details_converter.rb

```diff
@@ -0,0 +1,1270 @@
+# German bank details don't map directly to IBANs in the same way as in other
+# countries - each bank has idiosyncracies for translating its cusomers' bank
+# details. These idiosyncracies are described in a document from the Bundesbank:
+# http://www.bundesbank.de/Navigation/DE/Aufgaben/Unbarer_Zahlungsverkehr/IBAN_Regeln/iban_regeln.html
+
+module Ibandit
+  module GermanDetailsConverter
+    def self.rules
+      @rules ||= YAML.load_file(
+        File.expand_path('../../../data/german_iban_rules.yml', __FILE__)
+      )
+    end
+
+    def self.convert(opts)
+      # Fetch the relevant rule number. Default to '000000'
+      rule_num = rules.fetch(opts[:bank_code], {}).fetch(:iban_rule, '000000')
+
+      # Convert the bank details using the relevant rule
+      updated_bank_details = const_get("Rule#{rule_num}").new(
+        opts[:bank_code],
+        opts[:account_number]
+      ).converted_details
+
+      opts.merge(updated_bank_details)
+    end
+
+    ##############
+    # IBAN Rules #
+    ##############
+
+    class BaseRule
+      def initialize(bank_code, account_number)
+        @bank_code = bank_code
+        @account_number = account_number
+      end
+
+      def converted_details
+        raise NotImplementedError, 'Concrete RuleXXXXXX classes should ' \
+                                   'define a converted_details function'
+      end
+    end
+
+    module PseudoAccountNumberBehaviour
+      def converted_details
+        updated_account_number =
+          self.class.const_get('PSEUDO_ACCOUNT_NUMBER_MAPPING').fetch(
+            @account_number.rjust(10, '0'),
+            @account_number
+          )
+
+        { bank_code: @bank_code, account_number: updated_account_number }
+      end
+    end
+
+    class Rule000000 < BaseRule
+      def converted_details
+        { bank_code: @bank_code, account_number: @account_number }
+      end
+    end
+
+    class Rule000100 < BaseRule
+      def converted_details
+        msg = "Bank code #{@bank_code} is not used for payment transactions"
+        raise UnsupportedAccountDetails, msg
+      end
+    end
+
+    class Rule000200 < BaseRule
+      def converted_details
+        if @account_number.rjust(10, '0')[7] == '6' ||
+           @account_number.rjust(10, '0').slice(7, 2) == '86'
+          msg = 'Account does not support payment transactions'
+          raise UnsupportedAccountDetails, msg
+        else
+          { bank_code: @bank_code, account_number: @account_number }
+        end
+      end
+    end
+
+    class Rule000300 < BaseRule
+      def converted_details
+        if @account_number == '6161604670'
+          msg = 'Account does not support payment transactions'
+          raise UnsupportedAccountDetails, msg
+        else
+          { bank_code: @bank_code, account_number: @account_number }
+        end
+      end
+    end
+
+    class Rule000400 < BaseRule
+      include PseudoAccountNumberBehaviour
+
+      PSEUDO_ACCOUNT_NUMBER_MAPPING = {
```

Comment:
With the class instance variable approach this would then before `@pseudo_account_number_mapping = ...`